### PR TITLE
[TIR][CUDA] Add native FP8 support to codegen

### PR DESCRIPTION
**Pull request by csullivan**
Adds native FP8 type support for CUDA. The e4m3/e5m2 struct types provide explicit type conversions that target hardware native conversion ops. 

\* Conditionally run Storage and Compute legalization for targets that don't support FP8. This could be changed to only support conversion operators and do legalization on any compute operations other than builtin `wmma` calls. 

\* Implement support for `float16x4 (half4)` for use with `e4m3_float8x4 (__nv_fp8x4_e4m3)`

e.g.
```C++
#include <cuda_fp8.h>
using fp8_e4_t = __nv_fp8_e4m3;
using fp8_e4_2_t = __nv_fp8x2_e4m3;
using fp8_e4_4_t = __nv_fp8x4_e4m3;
using fp8_e5_t = __nv_fp8_e5m2;
using fp8_e5_2_t = __nv_fp8x2_e5m2;
using fp8_e5_4_t = __nv_fp8x4_e5m2;

struct __align__(4) half4 {
  __half x, y, z, w;
  __host__ __device__ half4() : x(__half(0)), y(__half(0)), z(__half(0)), w(__half(0)) {}
  __host__ __device__ half4(__half x, __half y, __half z, __half w) : x(x), y(y), z(z), w(w) {}
  __host__ __device__ explicit half4(const __nv_fp8x4_e4m3& fp8x4);
  __host__ __device__ explicit operator __nv_fp8x4_e4m3() const;
};
#endif


extern "C" __global__ void __launch_bounds__(32) add_kernel(fp8_e4_4_t* __restrict__ A, fp8_e4_4_t* __restrict__ B, fp8_e4_4_t* __restrict__ C) {
  half4 __1;
    half4 v_ = (half4)(A[((((int)blockIdx.x) * 32) + ((int)threadIdx.x))]);
    half4 v__1 = (half4)(B[((((int)blockIdx.x) * 32) + ((int)threadIdx.x))]);
    __1.x = (v_.x+v__1.x);
    __1.y = (v_.y+v__1.y);
    __1.z = (v_.z+v__1.z);
    __1.w = (v_.w+v__1.w);
  C[((((int)blockIdx.x) * 32) + ((int)threadIdx.x))] = (fp8_e4_4_t)(__1);
}
}
```

### CODE REVIEW

**Review by vinx13 — APPROVED** [review]

**Comment by tqchen** [comment]
please checkin on the ci issues, likely we need requires_cuda tag?

**Comment by tqchen** [comment]
@tvm-bot rerun

File: src/target/llvm/codegen_llvm.cc [hunk 1]
```diff
@@ -594,6 +596,7 @@ llvm::Type* CodeGenLLVM::DTypeToLLVMType(const DataType& dtype) const {
     return llvm::VectorType::get(etype, dtype.lanes());
 #endif
   } else {
+    ICHECK(etype != nullptr) << "No suitable llvm type found for dtype: " << dtype;
```
Comment (vinx13) [inline src/target/llvm/codegen_llvm.cc:599]: this breaks CI

Comment (csullivan) [reply src/target/llvm/codegen_llvm.cc]: Thank you @vinx13. I was trying to leave the code better than I found it, but I guess some downstream handling is relying on this behavior. I have removed the check.

**Pull request merged**